Implementation/57515 show a list of all projects in which a given custom field is activated

akabiru opened the pull request:
# Ticket
<!-- Provide the link to respective work package -->

Follows #16543 

https://community.openproject.org/wp/57515

<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->

# What are you trying to accomplish?
<!-- Provide a description of the changes. -->

List projects where a given custom field is enabled. In cases where the CF is enabled for all projects, show a notice of the same.

## Screenshots
<!-- Provide before/after screenshots, videos, or graphs for any visual changes; otherwise, remove this section -->

https://github.com/user-attachments/assets/99655ce6-5393-4d74-bb52-a1ecb5f22c80


# What approach did you choose and why?
<!-- This section is a place for you to describe your thought process in making these changes.
     List any tradeoffs you made to take on or pay down tech debt.
     Describe any alternative approaches you considered and why you discarded them. -->

This PR largely builds on the Project Attributes functionality - as a _"Project Attribute"_ is a special type of Custom Field known in the domain as "_ProjectCustomField_". The project queries are kept separate since the join table is different in both cases

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

akabiru committed:
[Op#5715]: Include custom fields and project attributes in available custom fields scope

akabiru committed:
[Op#5715]: Define `AvailableCustomFieldsProjectFilter`

akabiru committed:
[Op#57515]: Reuse custom fields view components

akabiru committed:
[Op#57515]: Set up first base view for custom field projects

akabiru committed:
[Op#57515]: avail custom field projects list

akabiru committed:
tests[Op#5715]: update test description

akabiru committed:
chore[Op#57515]: Remove unneeded template override

Kharonus commented on app/components/admin/custom_fields/custom_field_projects/row_component.rb:
❓ Why extracting row and table component logic to the `admin` namespace, if it is not used elsewhere? Why couldn't it stay at `settings` namespace?

Kharonus replied:
I think I found it, you are using it in the view template from `admin/custom_fields/custom_fields_projects/index.html.erb`.

akabiru replied:
Precisely. `{Admin::CustomFields::CustomFieldProjects,Settings::ProjectCustomFields}::{RowCompoent,TableComponent}` are reserved - only the inheritance is elevated to the primary model in this context which is a `CustomField` - a "ProjectAttribute" is only a special type of custom field identified as `ProjectCustomField` via an STI model.

akabiru replied:
https://github.com/opf/openproject/pull/16565/commits/f240ef707725f375f6cd2678e54c17e0b0d8d861

Kharonus commented on app/models/queries/projects/filters/available_project_custom_fields_filter.rb:
```diff
@@ -0,0 +1,70 @@
+  def apply_to(_query_scope)
+    case operator
+    when "="
+      super.with_available_custom_fields(values)
```
🟡 maybe I am missing part of the logic you have written/extracted here, but why exactly is this line not using `super.with_available_project_custom_fields(values)` in a filter name `AvailableProjectCustomFieldsFilter`?

akabiru replied:
The naming is confusing - I should rename this filter to `AvailableCustomFieldProjectsFilter` to clarify that we're querying against `CustomFieldsProject` which is the join table for all custom field types, except "ProjectCustomField" aka "Project Attributes" which uses a special join table `ProjectCustomFieldsProjectMapping`

Hence, the scopes in use are different as both filters query different tables. Good eye! 👍🏾

Kharonus reviewed (commented):
Not yet approving it, but also unsure, if the comment is hinting at something that needs change. :D

akabiru committed:
chore[Op#57515]: Redefine filter as `Filters::AvailableCustomFieldsProjectsFilter`

clarify that we're querying against `CustomFieldsProject` which is the join table for all custom field types, except "ProjectCustomField" aka "Project Attributes" which uses a special join table `ProjectCustomFieldsProjectMapping`

Kharonus reviewed (approved):
LGTM, fine work
